### PR TITLE
Moved to registry client for registration and heartbeats 

### LINKED ISSUE #9: Sidecar doesn't re-register with the Registry upon failed heartbeats
Currently, the sidecar registration code works by registering the service instance once, and then heartbeating periodically, while ignoring the response code from heartbeating:
https://github.com/amalgam8/sidecar/blob/master/register/svcregistration.go#L313

That logic doesn't take into account that the service instance might be expired by the registry (normally due to missed heartbeats), which results with 410 status code. In such case, the sidecar should re-register its instance with the registry.

Here's an example log from a sidecar client:
```
time="2016-06-26T08:16:34Z" level=info msg=Registering 
time="2016-06-26T08:16:34Z" level=info msg="Retrieved local IP" 
time="2016-06-26T08:16:34Z" level=info msg="Successfully registered with Registry" heartbeat_url="http://registry:8080/api/v1/instances/5f940f0ddee732bb/heartbeat" request_id= 
time="2016-06-26T08:16:34Z" level=info msg="Successfully registered with Registry" heartbeat_url="http://registry:8080/api/v1/instances/5f940f0ddee732bb/heartbeat" 
time="2016-06-26T08:16:34Z" level=info msg="Launching app 'python' with args 'reviews.py 9080 http://localhost:6379'" 
time="2016-06-26T19:17:53Z" level=error msg="Couldn't send heartbeat to Registry" err="Expected HTTP status 200 OK, got 410" 
time="2016-06-26T19:18:23Z" level=error msg="Couldn't send heartbeat to Registry" err="Expected HTTP status 200 OK, got 410" 
time="2016-06-26T19:18:53Z" level=error msg="Couldn't send heartbeat to Registry" err="Expected HTTP status 200 OK, got 410" 
time="2016-06-26T19:19:23Z" level=error msg="Couldn't send heartbeat to Registry" err="Expected HTTP status 200 OK, got 410" 
time="2016-06-26T19:19:53Z" level=error msg="Couldn't send heartbeat to Registry" err="Expected HTTP status 200 OK, got 410" 
time="2016-06-26T19:20:23Z" level=error msg="Couldn't send heartbeat to Registry" err="Expected HTTP status 200 OK, got 410" 
time="2016-06-26T19:20:53Z" level=error msg="Couldn't send heartbeat to Registry" err="Expected HTTP status 200 OK, got 410" 
time="2016-06-26T19:21:23Z" level=error msg="Couldn't send heartbeat to Registry" err="Expected HTTP status 200 OK, got 410" 
```

And the corresponding logs from the Registry server:
```
time="2016-06-26T19:17:38Z" level=debug msg="Instance ID 5f940f0ddee732bb is expired" module=STORE 
time="2016-06-26T19:17:53Z" level=warning msg="Failed to renew instance 5f940f0ddee732bb" error="2 - no such service instance (5f940f0ddee732bb)" module=AMALGAM8 namespace=local 
time="2016-06-26T19:17:53Z" level=info msg="172.17.0.9 PUT /api/v1/instances/5f940f0ddee732bb/heartbeat HTTP/1.1" apptype=service-discovery byteswritten=77 elapsedtime=888.397µs method=Renew module=ACCESS namespace=local protocol=Amalgam8 returncode=410 sd-request-id="dMueKNOuoT_1466968673737638441_7613" 
time="2016-06-26T19:18:23Z" level=warning msg="Failed to renew instance 5f940f0ddee732bb" error="2 - no such service instance (5f940f0ddee732bb)" module=AMALGAM8 namespace=local 
time="2016-06-26T19:18:23Z" level=info msg="172.17.0.9 PUT /api/v1/instances/5f940f0ddee732bb/heartbeat HTTP/1.1" apptype=service-discovery byteswritten=77 elapsedtime=918.349µs method=Renew module=ACCESS namespace=local protocol=Amalgam8 returncode=410 sd-request-id="dMueKNOuoT_1466968703737495452_7619" 
time="2016-06-26T19:18:53Z" level=warning msg="Failed to renew instance 5f940f0ddee732bb" error="2 - no such service instance (5f940f0ddee732bb)" module=AMALGAM8 namespace=local 
time="2016-06-26T19:18:53Z" level=info msg="172.17.0.9 PUT /api/v1/instances/5f940f0ddee732bb/heartbeat HTTP/1.1" apptype=service-discovery byteswritten=77 elapsedtime=1.242126ms method=Renew module=ACCESS namespace=local protocol=Amalgam8 returncode=410 sd-request-id="dMueKNOuoT_1466968733737278888_7625" 
time="2016-06-26T19:19:23Z" level=warning msg="Failed to renew instance 5f940f0ddee732bb" error="2 - no such service instance (5f940f0ddee732bb)" module=AMALGAM8 namespace=local 
time="2016-06-26T19:19:23Z" level=info msg="172.17.0.9 PUT /api/v1/instances/5f940f0ddee732bb/heartbeat HTTP/1.1" apptype=service-discovery byteswritten=77 elapsedtime=1.403054ms method=Renew module=ACCESS namespace=local protocol=Amalgam8 returncode=410 sd-request-id="dMueKNOuoT_1466968763737309842_7631" 
time="2016-06-26T19:19:53Z" level=warning msg="Failed to renew instance 5f940f0ddee732bb" error="2 - no such service instance (5f940f0ddee732bb)" module=AMALGAM8 namespace=local 
time="2016-06-26T19:19:53Z" level=info msg="172.17.0.9 PUT /api/v1/instances/5f940f0ddee732bb/heartbeat HTTP/1.1" apptype=service-discovery byteswritten=77 elapsedtime=943.806µs method=Renew module=ACCESS namespace=local protocol=Amalgam8 returncode=410 sd-request-id="dMueKNOuoT_1466968793737358977_7637" 
time="2016-06-26T19:20:23Z" level=warning msg="Failed to renew instance 5f940f0ddee732bb" error="2 - no such service instance (5f940f0ddee732bb)" module=AMALGAM8 namespace=local 
time="2016-06-26T19:20:23Z" level=info msg="172.17.0.9 PUT /api/v1/instances/5f940f0ddee732bb/heartbeat HTTP/1.1" apptype=service-discovery byteswritten=77 elapsedtime=1.812117ms method=Renew module=ACCESS namespace=local protocol=Amalgam8 returncode=410 sd-request-id="dMueKNOuoT_1466968823738416054_7643" 
time="2016-06-26T19:20:53Z" level=warning msg="Failed to renew instance 5f940f0ddee732bb" error="2 - no such service instance (5f940f0ddee732bb)" module=AMALGAM8 namespace=local 
time="2016-06-26T19:20:53Z" level=info msg="172.17.0.9 PUT /api/v1/instances/5f940f0ddee732bb/heartbeat HTTP/1.1" apptype=service-discovery byteswritten=77 elapsedtime=824.012µs method=Renew module=ACCESS namespace=local protocol=Amalgam8 returncode=410 sd-request-id="dMueKNOuoT_1466968853738303385_7649" 
time="2016-06-26T19:21:23Z" level=warning msg="Failed to renew instance 5f940f0ddee732bb" error="2 - no such service instance (5f940f0ddee732bb)" module=AMALGAM8 namespace=local 
time="2016-06-26T19:21:23Z" level=info msg="172.17.0.9 PUT /api/v1/instances/5f940f0ddee732bb/heartbeat HTTP/1.1" apptype=service-discovery byteswritten=77 elapsedtime=272.992µs method=Renew module=ACCESS namespace=local protocol=Amalgam8 returncode=410 sd-request-id="dMueKNOuoT_1466968883738054091_7655" 
```

### LINKED ISSUE #10: Utilize the client provided by the Registry
The Registry has added a new 'client' package, exposing a client interface/implementation:
https://github.com/amalgam8/registry/tree/master/client

The sidecar can leverage this client for its interaction with the Registry, instead of reimplementing this logic in its codebase.

### PR DESCRIPTION
**Pull request by ijsnellf**
Moved to using the client provided by registry for service registration and heartbeats. This change resolves #10 and resolves #9.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -102,7 +103,23 @@ func sidecarMain(conf config.Config) error {
...
 		}
 		logrus.Info("Registering")
-		register.DoServiceRegistrationAndHeartbeat(&conf, true)
+
+		agent, err := register.NewRegistrationAgent(register.RegistrationConfig{
+			Client: client.New(client.Config{
+				URL: conf.Registry.URL,
+				AuthToken: conf.Registry.Token,
+			}),
+			ServiceInstance: client.ServiceInstance{
```
Comment (zcahana) [inline main.go]: The ServiceInstance is missing the following information:
- Endpoint (conf.EndpointHost, conf.EndpointPort,or LocalIP())
- conf.ServiceVerion, that was passed-in as part of the metadata field (I guess this is what the controller is expecting?)

Also, the TTL value (60) has changed compared to the original code (45), and so has the heartbeats interval (20 vs 30). Both old & new values are sensible - just making sure this is intentional...

File: register/registration.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		return
+	}
+	agent.active = false
+
+	agent.stop <- struct{}{}
+}
+
+func (agent *RegistrationAgent) register() {
+	for {
+		registeredInstance, err := agent.config.Client.Register(agent.config.ServiceInstance)
```
Comment (zcahana) [inline register/registration.go]: We should probably add some logging here, for the following events:
1. Successful registration, including registered instance's ID
2. Failed heartbeats
3. Deregistration

File: main.go [hunk 3]
```diff
@@ -104,18 +104,24 @@ func sidecarMain(conf config.Config) error {
...
+		}
+
 		agent, err := register.NewRegistrationAgent(register.RegistrationConfig{
-			Client: client.New(client.Config{
-				URL: conf.Registry.URL,
-				AuthToken: conf.Registry.Token,
-			}),
-			ServiceInstance: client.ServiceInstance{
+			Client: registryClient,
+			ServiceInstance: &client.ServiceInstance{
```
Comment (zcahana) [inline main.go]: The ServiceInstance is missing the following information:

Endpoint (conf.EndpointHost, conf.EndpointPort,or LocalIP())
conf.ServiceVerion, that was passed-in as part of the metadata field (I guess this is what the controller is expecting?)
Also, the TTL value (60) has changed compared to the original code (45), and so has the heartbeats interval (20 vs 30). Both old & new values are sensible - just making sure this is intentional...

File: register/registration.go (hunk 2 above)
Comment (ijsnellf) [inline register/registration.go]: I agree. I added logging on failures, registration and de-registration.

I also noticed that some of the logic was bad --we were attempting to renew on a failed register and not attempting to re-register on a failed register. I fixed both issues.

File: main.go (hunk 3 above)
Comment (ijsnellf) [inline main.go]: @zcahana: this should be resolved in #14.

**Pull request merged**